### DISCUSSION TITLE
Best practice, if any, for workflows?

### DISCUSSION THREAD

**Discussion by squirrelchew** [jhuckaby/Cronicle#482] [General]
I've been hunting for a good automation system for a while now.  Cronicle seems flexible in the right ways, but the missing key here is the concept of workflows.

Cronicle seems to be built around scheduling specific tasks, and sure a task can call a second task, which can call a third, and so on, but this kind of setup is buried and nightmarish to support in a production setting where there are hundreds of jobs each with a half dozen steps.

Basically, I'm looking for the usability and flexibility of something like rundeck, but with the arbitrary job/task level concurrency and queuing of Cronicle.  Is what I'm looking for beyond the scope of Cronicle?

**Comment by jhuckaby** [comment]
Hey there!  Thank you for your interest in Cronicle.  Workflows are indeed coming!  It's one of the major features being introduced in Cronicle 2.0 (which is being renamed to Orchestra).  It is due for release this year (2022).

I have to be honest, this is the first I've heard of [rundeck](https://www.rundeck.com/), so I need to do some research and see what it can do, so I understand exactly what you mean by "usability and flexibility".  Orchestra Workflows will be an ordered "sequence" of jobs, which is implemented in the UI as a flexible list, with drag-drop reordering, and options you can set for each job.  Also, jobs can easily pass data and files to the next job in the list, and control the flow (skip, jump around).  Are these the kinds of features you are looking for?

I see that rundeck has a text-based workflow editor with an actual interpreted language for decision making and tree branching:

<img width="854" alt="Screen Shot 2022-03-18 at 9 20 44 AM" src="https://user-images.githubusercontent.com/700765/159043633-c34ec8a7-76a5-4bad-baca-8d28f48a2f19.png">

I have to say, this is **much** different than the workflow implementation in Orchestra.  I don't want my users to have to learn a new language to use the product.  While this offers great flexibility, it is also a huge leap in complexity, and will be a barrier for new users.

Orchestra's workflow implementation is much simpler -- the UI is basically a reorderable list of jobs to run.  However, the flexibility is there under the covers.  Each job can control the flow by sending commands back to the workflow controller.  An individual job can:

- Skip over jobs, or skip backwards
- Jump to any job in the list by its Event ID
- Retry itself
- Abort the entire workflow
- Send JSON data to the next job in the list
- Send HTML data to the workflow controller to display in the UI in real-time
- Upload entire files for any future job to get access to

You can also insert delays between jobs, and add decisions in the UI based on job outcomes.  Each job can also trigger actions like notifications and web hooks.  Oh, and each job in a workflow can itself be another workflow (workflowception).

You can also choose to run workflow jobs in parallel instead of in sequence, but you lose out on the chaining features in that case.  Parallel mode is mainly for "multiplexing" (running a single job on multiple servers in parallel).

Please let me know if these features will suffice for your needs, or if I need to implement something else that rundeck provides.

Thank you!

-- Joe

**Comment by jhuckaby** [comment]
**Update:** I dug further into some rundeck videos, and I found a more traditional workflow sequence editor here:

<img width="1049" alt="Screen Shot 2022-03-18 at 10 26 19 AM" src="https://user-images.githubusercontent.com/700765/159053018-890da626-d11c-4d15-9e62-71ab5f2284ac.png">

I guess the interpreted language editor is optional, and they also provide a job list.  The latter is more akin to what Orchestra will provide, although I must say my UI will be much prettier 😉

**Reply by mcfetz** [reply]
I am very excited about the new orchestra... 🤩

**Comment by squirrelchew** [comment]
Hooo boy, I've never jumped into the Rundeck DSL.  I primarily use shell commands within a workflow, as shown in your second comment.  Passing data from step to step is also a nice thing to have, but not anything I require (yet).

It sounds like the 2.0 release will absolutely be what I'm looking for, at least functionally speaking.  What's pictured in your second comment are the individual tasks in a workflow, and, if I'm understanding you correctly, Cronicle exclusively schedules and runs those constituent tasks until the 2.0 release where they will be able to be launched as a bundled sequence of tasks (perhaps configurable in a DAG, like OpsWise?).

From what I currently understand, tasks are written against plugins.  For my purposes, I almost exclusively use local shell commands, albeit they are "local" to the destination server.  Rundeck facilitates this as running on a "node" without needing anything actually installed (other than an SSH key) on the destination server.  A node is simply a configuration string for a destination server to execute things on using SSH.

I'm sure a Cronicle plugin could do effectively the same thing -- like a "Remote Shell Script" instead of "Shell Script".  Otherwise it could just, within the shell script, issue "ssh user@targetserver command" or some such.  Not ideal, but workable.  Otherwise, ultimately, it looks like Cronicle can have an agent on the destination server as a slave, and the scheduled task could simply target that particular agent.  This is honestly the correct way to do it, with the master server handling all the actual scheduling, but it's super nice not to have to install things on other servers.

The UI and job structuring remains a concern, albeit this is purely from my point of ignorance on basically all things Cronicle.  I went as far as sitting down and playing with the category-based arbitrary concurrency and queuing and that's about it.

One of the major, major boons of Rundeck is the approachability of the UI for a standard user.  Right now, my jobs are sorted out in a Context->Source->JobName style tree/folder view.  I find this extremely approachable for non-tech savvy users needing to utilize the automation system to launch jobs.

An example of this tree sorting would be CarType->CarCountry->CarMake->CarModel, so we'd have sedan, coupe, truck at the root level.  A user would log in and see just these three things.  They need to do something with coupes, so they expand coupe.  Now they see American, Asian, European.  They need to do something with an American car, so they expand American, and now they see Ford, Chevy, Dodge, etc.. you get the idea.  From a support/maintenance standpoint, I can be told "hey, something's wrong with the Ford Focus job" and I'd be able to quickly and easily navigate to that job.

"Categories" in Cronicle are a start to this, but I would like to see more powerful sorting of jobs to reduce visual noise to a user.  My system currently runs about 2,000 job executions per 24 hours across around 160 jobs that organize about 400 tasks -- these are very tiny numbers, for reference.  As it stands in Cronicle, I'd need to nebulously chain these 400 tasks into their 160 jobs, but if a user wanted to fire off one of those 160 jobs?  They'd need to (know how to) find the "first" job in that chain and kick that off.  No go.  I want to show them these digested categories -- further, I want to use these digested categories so I can maintain jobs that much easier!

That's how Rundeck does the UI.  Airflow is a half step ahead of where Cronicle currently is with its tagging/chiclet display for jobs.

I would be interested in any sort of available alpha/beta of the 2.0 release should it be available.  I'd love to do a proof of concept in our stack to see how it meshes with our needs.

**Reply by jhuckaby** [reply]
> if I'm understanding you correctly, Cronicle exclusively schedules and runs those constituent tasks until the 2.0 release where they will be able to be launched as a bundled sequence of tasks (perhaps configurable in a DAG, like OpsWise?).

You'll have to forgive me again, I don't know what a DAG is, nor have I heard of OpsWise.  I apologize -- I should know these things!  I will do some research.  However, if you are talking about automating job execution by way of a REST API, then both Cronicle and Orchestra provide this.  You can start any job or workflow with a simple HTTP GET to the master server (API Key for auth).

> it looks like Cronicle can have an agent on the destination server as a slave, and the scheduled task could simply target that particular agent. This is honestly the correct way to do it, with the master server handling all the actual scheduling, but it's super nice not to have to install things on other servers.

Yeah, that's the idea.  Cronicle and Orchestra greatly prefer that you have the remote agent running on all of your target servers, because then it can track the CPU/Memory usage of the job.  A remote SSH thing is easily doable through the use of a Plugin, but then you lose all the cool monitoring capabilities.  Orchestra takes this a big step further by providing process list snapshots, and a process tree of your job, not to mention full custom server monitoring during jobs.  But all this requires the agent to be running on the target server.

I will say this!  I've completely redesigned the "agent" in Orchestra, where it is now shipped as a single compiled binary file (for all Linux flavors + macOS), and can be installed with a single shell command, with no dependencies whatsoever.  So it's REALLY simple to throw up onto every server you have.

> An example of this tree sorting would be CarType->CarCountry->CarMake->CarModel, so we'd have sedan, coupe, truck at the root level. A user would log in and see just these three things. They need to do something with coupes, so they expand coupe. Now they see American, Asian, European. They need to do something with an American car, so they expand American, and now they see Ford, Chevy, Dodge, etc.. you get the idea. From a support/maintenance standpoint, I can be told "hey, something's wrong with the Ford Focus job" and I'd be able to quickly and easily navigate to that job.

Ah, I totally understand.  This is beyond what I had planned for Orchestra, but I will absolutely consider this.

Currently Orchestra has categories (like Cronicle), but also tags.  You can define any number of tags you want, and assign multiple tags to your events (a.k.a jobs), then easily search for those, so you only see the events you want in the list.

I do see the flexibility of rundeck's tree view, however:

<img width="801" alt="Screen Shot 2022-03-18 at 2 33 54 PM" src="https://user-images.githubusercontent.com/700765/159086042-f19fe00f-f905-4a37-96f7-0cf739b0216d.png">

I will look into implementing something like this.  Perhaps I can have "nested categories" or something.

Thank you for all these suggestions!  Rundeck looks like a great product.  I'm amazed I haven't heard of it until now.

**Reply by squirrelchew** [reply]
> You'll have to forgive me again, I don't know what a DAG is, nor have I heard of OpsWise. I apologize -- I should know these things! I will do some research. However, if you are talking about automating job execution by way of a REST API, then both Cronicle and Orchestra provide this. You can start any job or workflow with a simple HTTP GET to the master server (API Key for auth).

A DAG is a [Directed Acylic Graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph) -- a visualization of a workflow/job, if you'd like to think of it that way, that may only conditionally loop onto itself.  A lot of ETL work boils down into a DAG.

> I will say this! I've completely redesigned the "agent" in Orchestra, where it is now shipped as a single compiled binary file (for all Linux flavors + macOS), and can be installed with a single shell command, with no dependencies whatsoever. So it's REALLY simple to throw up onto every server you have.

Good man!  As far as agents go, this is the way to do it.  I like your minimalistic approach thus far.
> Ah, I totally understand. This is beyond what I had planned for Orchestra, but I will absolutely consider this.
> 
> Currently Orchestra has categories (like Cronicle), but also tags. You can define any number of tags you want, and assign multiple tags to your events (a.k.a jobs), then easily search for those, so you only see the events you want in the list.

My knee-jerk response to this is "but the user would then need to know what tags exist that they could search for."  It's definitely a case of me being used to things a certain way and preferring it, but for general use a tagging system becomes untenable very quickly.  Think of an enterprise deployment where there are literally thousands of jobs, and users have roles which mask out ones they straight up don't need to see, but they do need to be able to navigate what's available.  This is where Rundeck shines, organizationally speaking, if a bit archaic in setup.

> I will look into implementing something like this. Perhaps I can have "nested categories" or something.

Any kind of hierarchy like that should be very easy to convert into a tree view!  I'm sure there are node libraries out there that facilitate rendering out a hashmap (or ah.. an "object" in JS, is it?  sorry, novice node-ist here) into a tree.

> Thank you for all these suggestions! Rundeck looks like a great product. I'm amazed I haven't heard of it until now.

Rundeck is one of those "if you know, you know" applications.  Pagerduty's acquisition of it has definitely given it some visibility, but it's still shockingly unknown for all that it can do.

I'm more than happy to help!  I've been nonchalantly searching for a replacement for rundeck for weeks now and only just stumbled across Cronicle from some crappy "top free scheduling apps" article on some crappy website last night.

Unfortunately I'm not much of a Node guy, and even less of a UI/UX guy, but hopefully I can help in some way beyond feature requests!

**Reply by squirrelchew** [reply]
Oh and as far as opswise, it's a stupid expensive enterprise scheduling application.  Extremely heavy, but hilariously powerful.  Building a job was done in a DAG style view, complete with arbitrary resource management, incredible logging, and APIs to leverage.  It was very much not a self-service app though, like Rundeck poises itself to be.

I feel it is prudent to explain what I mean by "self service" -- my first foray into Rundeck was a company's prodops team spinning up an instance to allow other departments to do server management within specific parameters.  You'd log in and see a job to spin up an ephemeral environment, for example.  Launch it, it'd ask some parameters like your initials, how long you need the environment to be active, what configurations and applications it should adopt from a multi-select input field, etc..  All these parameters are coded by the administrator of Rundeck in this case, so you can railroad your users into very specific usage cases and take a nice, hands-off approach to what would otherwise be a ticket request needing to be read, interpreted, and fulfilled by a human running the very same build script on a server somewhere.

In terms of my uses, I have users allowed to log in to my rundeck instance and all they see are the reporting-specific jobs (which is any job with "Report" in the tree path, to circle back to that tree view I keep talking about).  They can fire a job and optionally change the recipient email list for that specific instance of the job.  No more calling me at 7AM to rerun a report! :)

**Reply by jhuckaby** [reply]
> Any kind of hierarchy like that should be very easy to convert into a tree view! I'm sure there are node libraries out there that facilitate rendering out a hashmap (or ah.. an "object" in JS, is it? sorry, novice node-ist here) into a tree.

Oh I know, I get it.  Rendering a tree view UI in the browser is not the problem.  I've coded many over the years.  It's a matter of changing the app design to accommodate this.  Not only the job list page, but also search, filtering, bookmarking, creating the hierarchy, and presenting all that in a clean way with good UI / UX.

Also, regarding the job list UI, Rundeck merely lists the job title on each row.  Orchestra's event view is a full-width table with many columns, showing all kinds of things.  I've already put many months into the design and implementation, with the assumption that categories are "flat".  It's just going to be a lot of work to change everything around to support nested categorization / tree view like this.

I will add it to the TODO list and start working on it, but I can't promise this for the initial release.  It's a large base change to an app that is more than 3/4 complete at this point.

> I'm more than happy to help! I've been nonchalantly searching for a replacement for rundeck for weeks now and only just stumbled across Cronicle from some crappy "top free scheduling apps" article on some crappy website last night.  Unfortunately I'm not much of a Node guy, and even less of a UI/UX guy, but hopefully I can help in some way beyond feature requests!

Thank you very much for the offer!  I am currently coding Orchestra offline in stealth mode, but after release the open source version will be up on a public GitHub repo like Cronicle is.

I should point out that a user by the name of @mikeTWC1984 has forked Cronicle, and created his own version with lots of new features, including workflows.  You might want to check that out, if you don't want to wait for Orchestra: https://github.com/cronicle-edge/cronicle-edge

**Reply by jhuckaby** [reply]
I totally get what you mean about self-service usage, and creating a "form" for users to fill out to run a job.  That is a feature I haven't considered yet, so I will add it to the TODO list as well.  Thanks!

**Reply by squirrelchew** [reply]
Right on.  I very much look forward to the Orchestra release -- where should I be watching for that?  I'll also give cronicle-edge a look-see, thanks for the pointer.

**Reply by jhuckaby** [reply]
> I very much look forward to the Orchestra release -- where should I be watching for that?

Thanks again!  I'll post a new discussion topic here, augment the Cronicle README with a notice at the top, and also tweet from the [CronicleApp](https://twitter.com/cronicleapp) Twitter account, as well as [my own](https://twitter.com/jhuckaby), when Orchestra drops.  I guess I should start a new Twitter account for Orchestra, too!  I'll work on that.

Oh, and Orchestra's official website will be: https://orchestra.dev/ (currently under construction).
